skada/_subspace.py

```diff
@@ -366,3 +373,281 @@ def TransferComponentAnalysis(
+        -------
+        self : object
+            Returns self.
+        """
+        X, sample_domain = check_X_domain(
+            X,
+            sample_domain,
+            allow_multi_source=True,
+            allow_multi_target=True,
+        )
+        X_source, X_target = source_target_split(X, sample_domain=sample_domain)
+
+        if self.n_components is None:
+            n_components = min(min(X_source.shape), min(X_target.shape))
+        else:
+            n_components = self.n_components
+        self.random_state_ = check_random_state(self.random_state)
+        self.X_source_ = X_source
+        self.X_target_ = X_target
+
+        n = X.shape[0]
+        source_mask = extract_source_indices(sample_domain)
+
+        H = np.identity(n) - 1 / n * np.ones((n, n))
+        K = self._get_kernel_matrix(X_source, X_target)
+        M = self._get_mmd_matrix(X_source.shape[0], X_target.shape[0], sample_domain)
+        M /= np.linalg.norm(M, ord="fro")
+        G = np.identity(n)
+
+        # minimization of the objective function
+        # \min_{A} tr(A^T K M K A) + tradeoff * (||A_s||_{2, 1} + ||A_t||_F^2)
+        # s.t. A^T K H K^T A = I
+        EPS_eigval = 1e-10
+        # We take something that will force at least one iteration since loss>=0:
+        last_loss = -2 * self.tol
+        for i in range(self.max_iter):
+            # update A
+            B = K @ M @ K.T + self.tradeoff * G
+            C = K @ H @ K.T
+            B = B + EPS_eigval * np.identity(n)
+            C = C + EPS_eigval * np.identity(n)
+            phi, A = scipy.linalg.eigh(B, C)
+            phi = phi + EPS_eigval
+            indices = np.argsort(phi)[:n_components]
+            A = A[:, indices]
+
+            # update G
+            A_norms = np.linalg.norm(A, axis=1)
+            G = np.zeros(n, dtype=np.float64)
+            # ||A_s||_{2, 1}
+            G[A_norms != 0] = 1 / (2 * A_norms[A_norms != 0] + EPS_eigval)
+            # ||A_t||_F^2
+            G[~source_mask] = 1
+            G = np.diag(G)
+
+            loss = np.trace(A.T @ K @ M @ K @ A)
+            # np.linalg.norm(A[source_mask], axis=1).sum()
+            reg = (
+                np.sum(np.linalg.norm(A[source_mask], axis=1))
+                + np.linalg.norm(A[~source_mask]) ** 2
+            )
+            loss_total = loss + self.tradeoff * reg
+            # print objective function and constraint satisfaction
+            if self.verbose:
+                print(
+                    f"iter {i}: loss={loss_total:.4f}, loss_mmd={loss:.4f}, "
+                    f"reg={reg:.4f}"
+                )
+                mat = A.T @ K @ H @ K.T @ A
+                cond = np.allclose(mat, np.identity(n_components))
+                print(
+                    f"Constraint satisfaction: {cond}, dist="
+                    f"{np.linalg.norm(mat - np.identity(n_components))}"
+                )
+
+            if np.abs(last_loss - loss_total) < self.tol:
+                break
```

Comment:
The criteria must be relative, e.g `np.abs((last_loss - loss_total) / last_loss)`.